### ISSUE TITLE
one vhd is only allowed to mount once in one VM? How about 1 ReadWrite + 1ReadOnly?

### ISSUE THREAD

**Issue by andyzhangx** [khenidak/dysk#33]
@khenidak 
I have successfully mounted one vhd as ReadWrite on one VM and then I tried to mount the same vhd as ReadOnly in the same VM, while it failed, error is `cannot mount /dev/dyskRxVg1SDC read-only`. 
So my quesiton is one vhd is only allowed to mount once in one VM? How about 1 ReadWrite + 1ReadOnly on one VM for the same vhd?

Some logs:
```
Tue Mar 13 15:02:10 UTC 2018 EXEC: mkdir -p /var/lib/kubelet/pods/758851e8-26cf-11e8-a48d-000d3afdbef1/volumes/azure~dysk/test
Tue Mar 13 15:02:10 UTC 2018 INF: /etc/kubernetes/volumeplugins/azure~dysk/dyskctl mount -a andymghyb186 -k ... -c dysks -p dysk03.vhd --auto-lease --break-lease --read-only
Tue Mar 13 15:02:10 UTC 2018 EXEC: /bin/mount  --read-only /dev/dyskRxVg1SDC /var/lib/kubelet/pods/758851e8-26cf-11e8-a48d-000d3afdbef1/volumes/azure~dysk/test
mount: cannot mount /dev/dyskRxVg1SDC read-only
```
```
azureuser@k8s-agentpool-87187153-0:/etc/kubernetes/volumeplugins/azure~dysk$ sudo ./dyskctl list
Type                            Name                            VHD                             SizeGB                          AccountName                     Path
R                               dyskRxVg1SDC                    Yes                             2                               andymghyb186                    /dysks/dysk03.vhd
R                               dyskeOsuH3EV                    Yes                             2                               andymghyb186                    /dysks/dysk03.vhd
R                               dyskfMSYLNff                    Yes                             2                               andymghyb186                    /dysks/dysk03.vhd
R                               dyska7eQcCRb                    Yes                             2                               andymghyb186                    /dysks/dysk03.vhd
R                               dyskCV2r7xsV                    Yes                             2                               andymghyb186                    /dysks/dysk03.vhd
R                               dyskNlOCxegZ                    Yes                             2                               andymghyb186                    /dysks/dysk03.vhd
RW                              dyskyeBuRgoN                    Yes                             2                               andymghyb186                    /dysks/dysk03.vhd
```

**Comment by andyzhangx** [comment]
Update:
for such case, I found it will hung when I try to unmount:
```
azureuser@k8s-agentpool-87187153-0:/etc/kubernetes/volumeplugins/azure~dysk$ sudo ./dyskctl unmount -d dyskyeBuRgoN





```

**Comment by khenidak** [comment]
Hi Andy - can you confirm you are using `head` for your tests? The readOnly many + 1readWrite is only at head

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by andyzhangx** [comment]
I am using this installation https://github.com/khenidak/dysk/blob/master/kubernetes/deployment/dysk-kubernetes-installer.yaml, it's using "khenidak/dysk-flexvol-installer:0.5", is it most updated?

**Comment by andyzhangx** [comment]
pls update all images to latest, old image is using `dysk~dysk` instead of `azure~dysk`, the new dysk flexvolume will be ready soon. It's almost done.

**Comment by khenidak** [comment]
so re the original issue (also applies on #31) the issue is the cli was updated to not use leases for readonly but the module wasn't because the yaml in `installer` file was pointing to an old hash.I have updated them. Can you please test?

**Comment by khenidak** [comment]
> pls update all images to latest, old image is using dysk~dysk instead of azure~dysk, the new dysk flexvolume will be ready soon. It's almost done.

Once it is done, i will upgrade the images.

**Comment by andyzhangx** [comment]
@khenidak 
I have tried latest(with commit `value: 03d39745144154e23f65d62234924d8a3676cb44`), 1 RW and 1 RO does not work, below are my repro:
On VM#1:
```
sudo dyskctl mount -a andymghyb186 -k "..." -c dysks -p dysk05.vhd --auto-lease --break-lease
sudo mkfs.ext4 /dev/dyskdnJ4WXbS
sudo mkdir ~/test
sudo mount /dev/dyskdnJ4WXbS ~/test
```

On VM#2:
```
sudo dyskctl mount -a andymghyb186 -k "..." -c dysks -p dysk05.vhd --auto-lease --break-lease --read-only 
sudo mkdir ~/test
sudo mount --read-only  /dev/dyskoBIM8utf ~/test
mount: cannot mount /dev/dyskoBIM8utf read-only
```

And I found that when I mount read-only on VM#2, VM#1 will be dead, I cannot ssh to VM#1, what I can do is restart VM#1.

**Comment by andyzhangx** [comment]
Update:
I have just verified that multiple ReadOnly mount works, while 1 ReadWrite mount and 1 ReadOnly mount does not work with the latest image.
And I have completed the new dysk flexvolume driver which should only support static provioning(no `auto-create`), I will send a new PR to this repo later:
https://github.com/andyzhangx/kubernetes-drivers/tree/master/flexvolume/dysk

**Comment by khenidak** [comment]
Thanks. wrt the original issue. Here are my findings so far:
1- I believe the 2nd `readOnly` mount broke the lease for the first `readWrite` mount. This is a bug. I will have to check.
2- more critical - breaking the lease should unmount the dysk gracefully without panics.

**Comment by ehotinger** [comment]
@khenidak what was the conclusion here, can we close this out or is there still investigation necessary?

**Comment by khenidak** [comment]
Still validating and creating a test for this. Should be closed by next Tuesday.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by khenidak** [comment]
Got it `--auto-lease --break-lease` should be ignored in read only scenarios. I have ran the following test
`
./dyskctl mount auto-create ...
#format and mount the disk
#get the the vhd name of the newly created disk and
./dyskctl mount ... --readonly (notice no break-lease or --auto-lease)
#mount the disk as readonly
mount -o ro,noload /dev/XXXX /mnt/r 
`
works as expected. 

I believe the problem in your case that the `readonly` call (on 2nd VM) had `--break-lease` and `--auto-lease` broke the lease for the first VM. I will review the code against it. 

In all cases a losing a disk should not hang the machine. and that is what i am going to keep my focus on

**Comment by andyzhangx** [comment]
@khenidak I always use `--auto-lease --break-lease` in dysk flexvol driver no matter it's readonly or not, is that ok? I think in dysk level, there should have some protection code for this kind of case?

**Comment by khenidak** [comment]
The cli should ignore these flags if dysk is attached in `read-only` mode. They should be removed from flexvol for clarity (not for functional correctness).

**Comment by khenidak** [comment]
the cli does ignore these these flags when mounted readonly. I have verified it again using mount `--auto-lease --break-lease --read-only`. This is at the head. However i do believe there was a mistake at build time where images are not aligned (old cli/new module). I will align them in prep for flex vol merge.

I will close the issue later today, re-open if there is a regression.

**Comment by andyzhangx** [comment]
@khenidak could you update all images, and I would do a final sanity test of dysk flexvolume driver on k8s. Per my last test, there is system hung issue in multiple read/write mode.

**Comment by khenidak** [comment]
all images are now aligned to `0.6` I have reviewed all the yaml spec against it. and will sanity check it on k8s cluster.
